src/types/shared.ts

```diff
@@ -1,7 +1,18 @@
+/**
+ * Clusters provide the client application a mechanism for categorizing
+ * annotations so that their drawn anchor highlights may be styled distinctively
+ * in the annotated document. An annotation can only belong to one cluster.
+ */
+export type HighlightCluster =
+  | 'other-content' // default cluster: content not belonging to the current user
+  | 'user-annotations' // An annotation belonging to the current user
+  | 'user-highlights'; // A highlight (highlights are private; they always belong to the current user)
+
 /**
  * Annotation properties not present on API objects, but added by the client
  */
 export type ClientAnnotationData = {
+  $cluster?: HighlightCluster;
```

Comment:
The use of a `$cluster` property on annotations works for the current highlight clusters, because an annotation's cluster never changes once the `Annotation` object has been loaded/created. If in future we add new clustering methods which result in the annotation changing clusters dynamically then we'll either need to modify all the annotation objects (which could have performance overhead from invalidating various cached data, or just cache invalidation issues if we forget to update `$cluster` in some context), or we might need some system of having layers of cluster assignments, with a mechanism to resolve which one is active.

Reply:
Let's talk about this with voices in our next project check-in, but yes, this is important and something I've been keeping in mind as I go here. Very broadly, I could foresee:

* Static clusters known to the app (like this feature). Annotations either don’t change which cluster they’re in once loaded/created, or changes to an individual annotation’s cluster would happen if the annotation were modified. Changes to a single annotation’s cluster in the sidebar could be communicated to the annotator[^1] but should allow an existing highlight to be updated (i.e. CSS class changed or something), not redrawn. And certainly not require re-anchoring.
* Dynamic clustering. We’ve talked about a possible “emphasis mode”, for example, in which, say, search results or other sets of annotations could be visually intensified while other highlights would visually recede. I believe we could handle this similarly to how we currently handle focusing an annotation: the sidebar could pass the annotator a set of tags; the annotator could toggle a CSS class for highlight elements corresponding to those tags. Modally-applied styling would disable other cluster styling and make any highlight elements without the emphasis class visually recede. CSS class toggling — which we’re already doing for focused highlights — not re-drawing anchors.

tl;dr: Unless I do something really wrong, we should never need to modify multiple annotation objects or re-draw anchor highlights, and an annotation's `$cluster` value should only change on create/load or modify/save. We do need to support the ability for an annotation's `$cluster` to change when the annotation itself is modified.

[^1]: In thinking through how to explain this, I realized there is a small bug I'll need to kick out of the current implementation of this feature: in the slightly uncommon but not rare situation that a user edits a highlight and adds body text, it becomes an annotation. We need a mechanism to a) update an annotation's `$cluster` appropriately on modify/save (probably save) in the sidebar and b) communicate that change to the annotator. I'm adding that to my queue.